### PR TITLE
[NU-1836] Add casting and conversions docs

### PR DESCRIPTION
**Pull request by lukasz-bigorajski**
## Checklist before merge
- [ ] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [ ] Code is cleaned from temporary changes and commented out lines
- [ ] Parts of the code that are not easy to understand are documented in the code
- [ ] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [ ] Verify that PR will be squashed during merge

## Summary by CodeRabbit

- **Documentation**
	- Enhanced clarity and detail in the SpEL (Spring Expression Language) documentation.
	- Added new sections on explicit conversions and a matrix for data type conversions.
	- Updated terminology from "cast" to "convert" for consistency.
	- Improved document structure and formatting for better readability.

### CODE REVIEW

File: docs/scenarios_authoring/Spel.md [hunk 1]
```diff
@@ -264,18 +264,89 @@ can be accessed in e.g. in following ways:
...
+List of built-in functions:
+- `is(className)`/`to(className)`/`toOrNull(className)`
+- `isBoolean`/`toBoolean`/`toBooleanOrNull`
+- `isLong`/`toLong`/`toLongOrNull`
+- `isDouble`/`toDouble`/`toDoubleOrNull`
+- `isBigDecimal`/`toBigDecimal`/`toBigDecimalOrNull`
+- `isList`/`toList`/`toListOrNull`
+- `isMap`/`toMap`/`toMapOrNull`
+
+Functions with the prefix `is` check whether a type can be converted or cast to the appropriate type. Functions with
```
Comment (arkadius) [inline docs/scenarios_authoring/Spel.md:285]: Let's change the order to can be casted or converted to emphasize how the chain looks like

File: docs/scenarios_authoring/Spel.md [hunk 2]
```diff
@@ -264,18 +264,89 @@ can be accessed in e.g. in following ways:
...
+the `to` prefix convert or cast a value to the desired type, and if the conversion fails, an exception is propagated
+further. Functions with the `toOrNull` prefix convert or cast a value to the desired type, and if conversion fails, a
+null value is returned.
+The `is`, `to` and `toOrNull` functions are slightly different in that they take a class name as a parameter. They first
```
Comment (arkadius) [inline docs/scenarios_authoring/Spel.md:289]: Let's emphasize that we are talking about these methods without target class is name, with target class as a parameter to make it more obvious that we changing a subject. Also, let's add a spacing between these paragraphs

File: docs/scenarios_authoring/Spel.md [hunk 3]
```diff
@@ -264,18 +264,89 @@ can be accessed in e.g. in following ways:
...
+attempt to cast a value to the specified class. If the cast fails and there is a defined conversion to that class, the
+conversion is applied. Below is the list of available conversion to classes:
```
Comment (arkadius) [inline docs/scenarios_authoring/Spel.md:291]: This list is basically the full list of types that are available in the Nussknacker so if we wanted to add it somewhere, it should be somewhere else. Also, I think that the list without explanation for each class doesn't add any value - user can explore it be their own experimenting with the tool. To sum up - let's remove it.

File: docs/scenarios_authoring/Spel.md [hunk 4]
```diff
@@ -264,18 +264,89 @@ can be accessed in e.g. in following ways:
...
+- UUID 
+- Short 
+- String 
+- ZoneId 
+- ZoneOffset
+
+Conversions only make sens between specific types. Below is a matrix which shows which types can be converted with each
+other:
+
+| FROM TYPE (column) \ TO TYPE (row) | BigDecimal | BigInteger | Boolean | Byte | Charset | ChronoLocalDate | ChronoLocalDateTime | Currency | Double | Float | Integer | List | Locale | LocalDate | LocalDateTime | LocalTime | Long | Map | Unknown | UUID | Short | String | ZoneId | ZoneOffset | All existing types |
```
Comment (arkadius) [inline docs/scenarios_authoring/Spel.md:319]: Nice, maybe let's use:
:heavy_check_mark: :  - possible
:x: - not possible
:exclamation: - potentially failing
See https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/types/#casting
At first glance I thought that it is a a matrix of possible conversions, but than I realize that it is a matrix of not possible conversions.

File: docs/scenarios_authoring/Spel.md [hunk 5]
```diff
@@ -264,18 +264,89 @@ can be accessed in e.g. in following ways:
...
+- LocalTime 
+- Long 
+- Map 
+- UUID 
+- Short 
+- String 
+- ZoneId 
+- ZoneOffset
+
+Conversions only make sens between specific types. Below is a matrix which shows which types can be converted with each
```
Comment (arkadius) [inline docs/scenarios_authoring/Spel.md:316]: sens -> sense.
I think that is worth mention that we are not showing conversions that have no sense in suggestions mechanism

**Review by ZbyszekMM — CHANGES_REQUESTED** [review]

File: docs/scenarios_authoring/Spel.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukasz-bigorajski) [reply docs/scenarios_authoring/Spel.md]: Changed to 'canBe'

Comment (lukasz-bigorajski) [reply docs/scenarios_authoring/Spel.md]: Done

File: docs/scenarios_authoring/Spel.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukasz-bigorajski) [reply docs/scenarios_authoring/Spel.md]: Done

Comment (lukasz-bigorajski) [reply docs/scenarios_authoring/Spel.md]: Done

File: docs/scenarios_authoring/Spel.md [hunk 6]
```diff
@@ -264,18 +264,89 @@ can be accessed in e.g. in following ways:
...
+Functions with the prefix `is` check whether a type can be converted or cast to the appropriate type. Functions with
+the `to` prefix convert or cast a value to the desired type, and if the conversion fails, an exception is propagated
+further. Functions with the `toOrNull` prefix convert or cast a value to the desired type, and if conversion fails, a
+null value is returned.
+The `is`, `to` and `toOrNull` functions are slightly different in that they take a class name as a parameter. They first
+attempt to cast a value to the specified class. If the cast fails and there is a defined conversion to that class, the
+conversion is applied. Below is the list of available conversion to classes:
+- BigDecimal 
+- BigInteger 
+- Boolean
```
[the remark this answers is not in the document]
Comment (lukasz-bigorajski) [reply docs/scenarios_authoring/Spel.md]: I rephrased it a little. I hope it's better or if it's not we will try to fix it.

File: docs/scenarios_authoring/Spel.md (hunk 4 above)
Comment (lukasz-bigorajski) [reply docs/scenarios_authoring/Spel.md]: Done

File: docs/scenarios_authoring/Spel.md (hunk 5 above)
Comment (lukasz-bigorajski) [reply docs/scenarios_authoring/Spel.md]: Done

File: docs/scenarios_authoring/Spel.md [hunk 7]
```diff
@@ -264,18 +264,96 @@ can be accessed in e.g. in following ways:
 * `#exampleObjects['someNestedObject']['someFieldInNestedObject']`
 * `#exampleObjects['someArrayWithObjects'][0]['someFieldInObjectInArray']`
 
-Every unknown accessed field/element will produce `Unknown` data type which can be further navigated or [cast](#Casting) to a desired type.
+Every unknown accessed field/element will produce `Unknown` data type, which can be further navigated or [converted](#type-conversions) to a desired type.
 
 ### Type conversions
 
-It is possible to convert from a type to another type and this can be done by implicit and explicit conversion.
+It is possible to convert or cast from a type to another type and this can be done by implicit and explicit conversion.
```
Comment (arkadius) [inline docs/scenarios_authoring/Spel.md:271]: Here is still the opposite order of convert and cast words

File: docs/scenarios_authoring/Spel.md [hunk 8]
```diff
@@ -264,18 +264,96 @@ can be accessed in e.g. in following ways:
...
+- `canBeLong`/`toLong`/`toLongOrNull`
+- `canBeDouble`/`toDouble`/`toDoubleOrNull`
+- `canBeBigDecimal`/`toBigDecimal`/`toBigDecimalOrNull`
+- `canBeList`/`toList`/`toListOrNull`
+- `canBeMap`/`toMap`/`toMapOrNull`
+
+The aforementioned functions first attempt to cast a value to the specified class. If the cast fails and there is a 
+defined conversion to that class, the conversion is applied.
+The `canBe`, `to` and `toOrNull` functions take the name of target class as a parameter, in contrast to, for
+example, `canBeLong` which has the name of target class in the function name.
```
Comment (arkadius) [inline docs/scenarios_authoring/Spel.md:288]: Let's explain why we have both `canBe` and these variants with target class in the function name by adding some addition to this sentence. For example ... and is the shortcut fro `canBe('Long')`

File: docs/scenarios_authoring/Spel.md [hunk 9]
```diff
@@ -264,18 +264,96 @@ can be accessed in e.g. in following ways:
...
+Examples of conversions:
+
+| Expression                 | Result | Type    |
+|----------------------------|--------|---------|
+| `"123".canBeDouble`        | true   | Boolean |
+| `"123".toDouble`           | 123.0  | Double  |
+| `"abc".toDoubleOrNull`     | null   | Double  |
+| `"123".canBe('Double')`    | true   | Boolean |
+| `"123".to('Double')`       | 123.0  | Double  |
+| `"abc".toOrNull('Double')` | null   | Double  |
```
Comment (arkadius) [inline docs/scenarios_authoring/Spel.md:304]: Let's add an example that is failing in runtime

File: docs/scenarios_authoring/Spel.md (hunk 4 above)
Comment (arkadius) [reply docs/scenarios_authoring/Spel.md]: :+1: 
One more note: It is more natural to keep source type in the row and the target type in the column.
Thanks to that, it is easier to find the correct row and after that swipe in to correct column so while reading the sentence "Can I cast <source type> to <target type>" you will look at row to find source type and to column to find the target type.

File: docs/scenarios_authoring/Spel.md (hunk 9 above)
Comment (arkadius) [reply docs/scenarios_authoring/Spel.md:304]: Also, let's add an example of conversion of some list to map using projection mechanism as it is tricky. For example:
```
{12, 23, 34}.![{key: 'key_' + #this, value: #this}]
```

**Pull request merged**